---
title: "SnRNA-sequencing analysis of human perivascular adipose tissue (PVAT)"
author: "Hannah M. L. Bazin"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
---

In this vignette, I analyse three datasets of single-nucleus RNA sequencing of human perivascular adipose tissue (PVAT). There are 12’657, 6’084 and 6’456 cells in three samples respectively that were sequenced on the Illumina NovaSeq 6000. The raw data can be found on the Gene Expression Omnibus website [at this link](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE164528). I am building on code from Holly A. R. Giles.

# Set up

```{r}

# Set random seed for reproducibility
set.seed(42)

```

## Load libraries

```{r}

library(GEOquery)
library(Seurat)
library(dplyr)
library(patchwork)
library(ggplot2)
library(ggpubr)
library(tidyverse)

```

## Set working directory

The working directory can be set by running the following command in the terminal: `setwd("/Users/hannahbazin/Desktop/Cambridge/Academics/Han_Lab/MPhil/mphil-project")`

## Load data

The data stems from single-nucleus RNA sequencing experiments of human perivascular adipose tissue from three different healthy donors, sequenced with Illumina NovaSeq 6000. The raw data can be found [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE164528). It was first published by Angueira et al. on 12 April 2021 in Nature Metabolism in the paper titled "Defining the lineage of thermogenic perivascular adipose tissue" ([link](https://pubmed.ncbi.nlm.nih.gov/33846639/)).

We read in the data using the `Read10X()` function which takes in three files (barcodes.tsv, genes.tsv, and matrix.mtx) and returns a UMI sparse count matrix representing the gene expression data. This function can be used because the data was generated using CellRanger.

```{r}

# Read in the 10X data
human1_data <- Read10X(data.dir = "data/GSE164528_RAW/GSM5068996_hPVAT1/")
human2_data <- Read10X(data.dir = "data/GSE164528_RAW/GSM5068997_hPVAT2/")
human3_data <- Read10X(data.dir = "data/GSE164528_RAW/GSM5068998_hPVAT3/")

```

We create Seurat objects from the raw data using `CreateSeuratObject()`. A Seurat object contains both the data (e.g. counts matrix) and the analysis (e.g. clustering results). The parameters `min.cell` and `min.features` allow for initial filtering: we remove genes expressed in less than three cells (sparse expression), as well as cells with fewer than 200 detected genes.

```{r}

# Create Seurat objects for each sample
human1 <- CreateSeuratObject(counts = human1_data, project = "GSE164528", min.cells = 3, min.features = 200)
human2 <- CreateSeuratObject(counts = human2_data, project = "GSE164528", min.cells = 3, min.features = 200)
human3 <- CreateSeuratObject(counts = human3_data, project = "GSE164528", min.cells = 3, min.features = 200)

# See number of cells in each sample
print(paste("Before QC, human1 sample contains", ncol(human1), "cells."))
print(paste("Before QC, human2 sample contains", ncol(human2), "cells."))
print(paste("Before QC, human3 sample contains", ncol(human3), "cells."))

```

# Pre-processing workflow

## Visualise QC metrics and select cells

### Add mitochondrial reads

We add the percentage of mitochondrial genes to the metadata for all three samples. This is an important Quality Control metric. In single nucleus RNA sequencing data, as opposed to those from single cell RNA sequencing, we expect close to no low mitochondrial gene expression. We need to filter out the samples with high mitochondrial gene expression, because this might indicate that the mitochondria attached to the nucleus during isolation.

The `[[` operator can add columns to object metadata. First, we store the mitochondrial QC stat here. The `^` character ensures it matches only the start of a string.

```{r}

# Add percentage of mitochondrial genes
human1[["percent.mt"]] <- PercentageFeatureSet(human1, pattern = "^MT-")
human2[["percent.mt"]] <- PercentageFeatureSet(human2, pattern = "^MT-")
human3[["percent.mt"]] <- PercentageFeatureSet(human3, pattern = "^MT-")

```

### Determine cut-off

We visualise Quality Control metrics for each of the samples and determine the cut-off for filtering the cells based on these plots.

```{r}

# Show QC metrics for the first 5 cells of each sample
head(human1@meta.data, 5)
head(human2@meta.data, 5)
head(human3@meta.data, 5)

```

```{r fig.height=7, fig.width=13, warning = FALSE}

# Plot QC metrics for each sample
VlnPlot(human1, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(human2, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
VlnPlot(human3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

```

```{r fig.height=7, fig.width=11}

# Plot FeatureScatter plots for each sample
plot1 <- FeatureScatter(human1, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(human1, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

plot1 <- FeatureScatter(human2, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(human2, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

plot1 <- FeatureScatter(human3, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(human3, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

### Filter cells

The filtering metrics are:

-   `nFeature_RNA`: number of non-zero expressed genes in one cell, thus indicates the diversity of gene expression within a cell. Low values suggest low-quality cells or empty droplets; high values may indicate doublets or multiplets.
-   `nCount_RNA`: total RNA count per cell, reflecting transcriptional activity. Low values suggest low-quality cells; high values may indicate doublets or multiplets.
-   `percent.mt`: percentage of mitochondrial genes in one cell. High values indicate bad quality of cells.

From the violin plots, we identify the filtering thresholds and select cells as follows:

-   Filter cells that have unique feature counts over 6'000 or below 1'000.
-   Filter cells that have RNA counts over 40'000 and below 500.
-   Filter cells that have \>1% mitochondrial counts.

We use the same cut-off values in all three samples for consistency.

```{r}

# Filter cells for QC
human1_filtered <- subset(human1, subset = nFeature_RNA > 1000 & nFeature_RNA < 6000 & nCount_RNA > 500 & nCount_RNA < 40000 & percent.mt < 1)
human2_filtered <- subset(human2, subset = nFeature_RNA > 1000 & nFeature_RNA < 6000 & nCount_RNA > 500 & nCount_RNA < 40000 & percent.mt < 1)
human3_filtered <- subset(human3, subset = nFeature_RNA > 1000 & nFeature_RNA < 6000 & nCount_RNA > 500 & nCount_RNA < 40000 & percent.mt < 1)

# See number of cells in each sample after QC
print(paste("After QC, human1 sample contains", ncol(human1_filtered), "cells."))
print(paste("After QC, human2 sample contains", ncol(human2_filtered), "cells."))
print(paste("After QC, human3 sample contains", ncol(human3_filtered), "cells."))

```

# Normalise data and find variable features

We normalise the data using a global-scaling method called "LogNormalize". This process normalises the feature expression levels for each cell by dividing by the total expression, multiplies this by a scale factor (10'000 by default), and applies a log transformation to the result. Normalised values are stored in `human1_filtered[["RNA"]]@data` for each of the three samples.

We also identify features that have high cell-to-cell variation, these are likely to have the most biological significance. The function returns 2'000 features per dataset by default.

```{r}

# Add metadata to account for sample origin
human1_filtered$sampleType <- "GSM5068996"
human2_filtered$sampleType <- "GSM5068997"
human3_filtered$sampleType <- "GSM5068998"

lib_list_filtered <- list(human1_filtered, human2_filtered, human3_filtered)

# Normalise data and find variable features
lib_list_filtered <- lapply(X = lib_list_filtered, FUN = function(seurat_object) {
  
  seurat_object <- NormalizeData(seurat_object)
  seurat_object <- FindVariableFeatures(seurat_object, selection.method = "vst", nfeatures = 2000)
  
})

# Select features that are repeatedly variable across datasets for integration
features <- SelectIntegrationFeatures(object.list = lib_list_filtered)

```

We visualise the variable features.

```{r fig.height=5, fig.width=11, warning=FALSE}

lapply(1:length(lib_list_filtered), function(i) {
 
   x <- lib_list_filtered[[i]]
  
  # Identify the 10 most highly variable genes
  top10 <- head(VariableFeatures(x), 10)

  # Plot variable features with and without labels
  plot1 <- VariableFeaturePlot(x)
  plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
  plot1 + plot2

})

```

# Integrate data

Integration of data from the three different donors is an important step in this snRNA-seq workflow. This ensures that cells cluster based on biological state and not on sample origin. Details can be found in [this vignette](https://satijalab.org/seurat/archive/v4.3/integration_introduction). I am building on code from this vignette and from Holly A. R. Giles.

## Find integration anchors

```{r}

# Find integration anchors
  # This is a time-consuming step, save the object

if (file.exists("data/analysis/anchors.RData")) {
  
  message("Loading existing anchors file.")
  
  load(file = "data/analysis/anchors.RData")
  
} else {
  
  message("Missing anchors file. Finding integration anchors.")
  
  anchors <- FindIntegrationAnchors(object.list = lib_list_filtered, anchor.features = features)
  
  save(anchors, file = "data/analysis/anchors.RData")

}

```

## Perform integration

```{r}

# Use anchors to generate an integrated data assay
  # This is a time-consuming step, save the object
  # Delete objects before running if new changes to code

if (file.exists("data/analysis/libs_combined.RData")) {
  
  load(file = "data/analysis/libs_combined.RData")
  
} else {
  
  libs_combined <- IntegrateData(anchorset = anchors)
  
  save(libs_combined, file = "data/analysis/libs_combined.RData")
  
}

# Check dimensions of libs_combined
libs_combined

```

# Scale data

We scale the data so that highly expressed genes do not dominate in the downstream analyses. The `ScaleData()` function ensures the mean expression across cells becomes 0, and the variance becomes 1. We also regress out unwanted sources of variation such as mitochondrial reads.

```{r}

# Ensure default assay is integrated data
DefaultAssay(libs_combined) <- "integrated"

# Scale data and regress out unwanted variation due to mitochondrial reads
libs_combined <- ScaleData(libs_combined, vars.to.regress = 'percent.mt')

```

# Linear dimensionality reduction

We perform PCA on the scaled data with the variable features as input.

## Run PCA

```{r}

# By default, only the previously determined variable features are used as input.

libs_combined_filt <- RunPCA(libs_combined, npcs = 30, seed.use = 42)

```

## Determine dimensionality of the dataset

We plot an elbow plot to determine the dimensionality of the dataset. From this plot we can see that the line starts to flatten significantly around PC 15. Therefore in the downstream analysis we keep 15 dimensions of the dataset.

```{r cache = FALSE}

elbow_plot <- ElbowPlot(libs_combined_filt, ndims = 25)

# Save elbow plot as PDF
if (!dir.exists("results/humanPVATsn")) dir.create("results/humanPVATsn")

pdf("results/humanPVATsn/elbow_plot.pdf", width = 8, height = 6)
print(elbow_plot)
invisible(dev.off())

# Display elbow plot
elbow_plot

```

## Visualise PCA results

We plot the PCA by sample origin to verify that the integration worked successfully. We also check that there is no clear pattern in the percentage of mitochondrial genes and that there is no excessive percentage of such genes in our data.

```{r}

DimPlot(libs_combined_filt, reduction = "pca", 
              group.by = "sampleType")

FeaturePlot(object = libs_combined_filt, reduction = "pca", features = "percent.mt")

```

## Run and visualise UMAP

We run and plot the UMAP dimensionality reduction, first coloured by sample origin to verify that the integration worked successfully, and then by individual sample origin.
Note: human1 is a healthy weight, human2 is overweight, and human3 is obese.

```{r message = FALSE, cache = FALSE}

libs_combined_filt <- RunUMAP(libs_combined_filt, reduction = "pca", dims = 1:15, seed.use = 42)

# Plot UMAP by sample type
p1 <- DimPlot(libs_combined_filt,
              reduction = "umap",
              group.by = "sampleType") +
  ggtitle("UMAP coloured by sample origin")

p2 <- DimPlot(libs_combined_filt,
              reduction = "umap",
              split.by = "sampleType",
              group.by = "sampleType",
              ncol = 3) +
  ggtitle("UMAP of all samples")

# Save plots as PDFs
if (!dir.exists("results/humanPVATsn")) dir.create("results/humanPVATsn")

pdf("results/humanPVATsn/UMAP_by_sample_origin.pdf", width = 8, height = 6)
print(p1)
invisible(dev.off())

pdf("results/humanPVATsn/UMAP_all_samples.pdf", width = 12, height = 6)
print(p2)
invisible(dev.off())

# Display plots
p1
p2

```

The percentage of mitochondrial gene counts seems well distributed across the clusters, therefore we can continue with the downstream analysis.

```{r}

mito_plot <- FeaturePlot(object = libs_combined_filt, reduction = "umap", features = "percent.mt") +
  ggtitle("UMAP coloured by percentage of mitochondrial genes")

pdf("results/humanPVATsn/UMAP_mt_percent.pdf", width = 12, height = 6)
print(mito_plot)
invisible(dev.off())

# Display plots
mito_plot

```

# Find neighbors and clusters

We adapt the resolution to create an approximate number of clusters we expect based on papers of reference (such as Angueira et al., 2021) - i.e. around 13 clusters.

```{r}

libs_combined_filt <- FindNeighbors(libs_combined_filt, reduction = "pca", dims = 1:15)

libs_combined_filt <- FindClusters(libs_combined_filt, resolution = 0.35, random.seed = 42)

# Look at cluster IDs of the first 5 cells
head(Idents(libs_combined_filt), 5)

```

## Visualise clusters on UMAP

We visualise the clusters based on cluster number.

```{r}

# Plot UMAP with cluster numbers
umap_with_numbers <- DimPlot(libs_combined_filt,
                            reduction = "umap",
                            label = TRUE) +
  ggtitle("UMAP with Cluster Numbers")

# Save plot
pdf("results/humanPVATsn/UMAP_with_cluster_numbers.pdf", width = 12, height = 6)
print(umap_with_numbers)
invisible(dev.off())

# Show plot
umap_with_numbers

```

```{r}
# Plot UMAP with cluster numbers, split by sample
umap_with_numbers_by_sample <- DimPlot(libs_combined_filt,
                            reduction = "umap",
                            split.by = "sampleType",
                            label = TRUE) +
  ggtitle("UMAP with Cluster Numbers Split by Sample")

# Save plot
pdf("results/humanPVATsn/UMAP_with_cluster_numbers_by_sample.pdf", width = 12, height = 6)
print(umap_with_numbers_by_sample)
invisible(dev.off())

# Show plot
umap_with_numbers_by_sample
```

# Find markers

Before differential gene expression analysis, we return to the RNA assay instead of the integrated one. The function `FindAllMarkers()` finds markers (differentially expressed genes) for each of the clusters in a dataset. By default, it uses a Wilcoxon Rank Sum test to identify differentially expressed genes between two groups. The parameters are set as follows:

- `min.pct = 0.3` only tests genes detected in 30% of either of the two populations being compared. This speeds up the function and allows us to focus on more biologically relevant genes.
- `logfc.threshold = 0.3` limits testing to genes that show at least 0.3-fold difference (log-scale) between the two groups (removes weaker signals).


```{r}

# Return to RNA assay as default for DGEA
DefaultAssay(libs_combined_filt) <- "RNA"

# Find markers
  # This is a time-consuming step, save the object
if(file.exists("data/analysis/cluster_markers.RData")){
  
  load(file = "data/analysis/cluster_markers.RData")

  }else{
    
    markers <- FindAllMarkers(libs_combined_filt, min.pct = 0.3, logfc.threshold = 0.3)
    
    save(markers, file = "data/analysis/cluster_markers.RData")

  }

# Show the top ten markers in each cluster
top8 <- markers %>% group_by(cluster) %>% slice_max(n = 8, order_by = avg_log2FC)

top10 <- markers %>% group_by(cluster) %>% slice_max(n = 10, order_by = avg_log2FC)

top20 <- markers %>% group_by(cluster) %>% slice_max(n = 20, order_by = avg_log2FC)

top10

```

We generate a table of cluster markers for reference.

```{r warning = FALSE}

# Determine number of clusters
num_clusters <- max(as.numeric(as.character(libs_combined_filt@meta.data$seurat_clusters)))

lapply(0:num_clusters, function(i) {
  
  # Select relevant information
  tab <- filter(markers, cluster == i) %>% 
    select(gene, p_val, p_val_adj, avg_log2FC, cluster)
  
  # Export as csv for Excel
  write.csv(tab, file = paste0("results/humanPVATsn/markersCluster", i, ".csv"))
  
  # Print as datatable
  DT::datatable(tab)
  
})

```

## Plot markers

We plot violin plots for the top 8 marker genes for each cluster.

```{r fig.height=20, fig.width=15}

# Ensure that RNA assay is used and not integrated assay
DefaultAssay(libs_combined_filt) <- "RNA"

violinPlots <-
  lapply(0:num_clusters, function(i) {
    
    # Get a list of genes
    features_to_plot <- filter(top8, cluster == i)$gene
    
    # Get a list of violin plots, 8 per cluster
    violinList <- VlnPlot(libs_combined_filt, features = features_to_plot)
    
    # Plot a set of 8 violin plots for each cluster
    violinList[[1]] + violinList[[2]] + violinList[[3]] + 
      violinList[[4]] + violinList[[5]] + violinList[[6]] + 
      violinList[[7]] + violinList[[8]] +
      # Add cluster number as title
      plot_annotation(title = paste("Cluster", i, sep = " "),
                      theme = theme(plot.title = element_text(size = 18, face = "bold")))

  })

violinPlots

```

## Plot dot plot to identify clusters

We generate different dot plots to identify which clusters correspond to which cell types with PVAT.


```{r}

DotPlot(libs_combined_filt,
        features = c(

          # Early markers
          "DCN", "FBLN1", "COL3A1", "COL1A1", "COL1A2", "FSTL1", "GSN", "FBLN5", "MFAP5", "ADAMTSL3", "IGFBP6", "SEMA3C", "CREB5"
          
          )) +
  
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  ggtitle("Early pre-adipocyte marker expression")

```

```{r}

DotPlot(libs_combined_filt,
        features = c(

          # Intermediate markers
          "COL15A1", "APOD", "EBF1"
          
          )) +
  
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  ggtitle("Intermediate pre-adipocyte marker expression")

```

```{r}

DotPlot(libs_combined_filt,
        features = c(

          # Pre-adipocyte markers
          "PDGFRA", "LAMA2", "FBLN1", "FBN1", "FSTL1", "NEGR1", "EBF2", "DCN", "MGP", "PPARG"
          
          )) +
  
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  ggtitle("Literature pre-adipocyte markers")

```

```{r fig.height=7, fig.width=10}

DotPlot(libs_combined_filt,
        features = c(
          # Early markers
          "DCN", "FBLN1", "COL3A1", "COL1A1", "COL1A2", "FSTL1", "GSN", "FBLN5", "MFAP5", "ADAMTSL3", "IGFBP6", "SEMA3C", "CREB5",
      
          # Intermediate markers
          "COL15A1", "APOD", "EBF1",
          
          # Pre-adipocyte
          "NEGR1","EBF2","MGP", "PPARG",
          # "DCN"
          
          # Mature adipocyte
          "CD3D","GHR","CIDEC", "PLIN1", "CIDEA", "ADIPOQ", "LPL", "LIPE", "UCP1",
          
          # T cell
          "PTPRC","CD3G",
          
          # Macrophages (until CD69)
          "FRMD4B","CD68","ITGAM","LPAR1","SOX9",
         
           # Endothelial
          "VWF", "PECAM1", "ANO2","FLT4",
          
          # SMC
          "MYH11", "MYO1D", "RYR2", "ACTA2",
          
          # Circadian rhythm
          "CRY2", "CSNK1E",
          
          # Other
          "CD14","FCGR3A",
          
          # Other
          "BCL11A","MS4A1","CD79A","IGHM",

          # Other
          "LAMP3","ITGAX","VIM","CCR7","SELL","CREM","CD69", "GATA6",
          
          "DPP4", "ICAM1"
          
          )) +
  
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

```


This dot plot contains the genes identified as important marker genes for the cell types identified in these clusters. They are listed in order of cell types.

```{r fig.height=7, fig.width=10}

annot_dot_plot <- DotPlot(libs_combined_filt,
                          cols = "PiYG",
        features = c(
          
          ## Adipocyte-related
          
          # Pre-adipocytes
          "NEGR1", "EBF2", "DCN", "MGP", "PDGFRA", "FBLN1", "FBLN2", "FBLN5", "COL3A1", "COL1A1", "COL1A2", "FSTL1", "GSN", "ADAMTSL3", "CREB5", "PPARG",
          
          # Mature brown adipocytes
          "ADIPOQ", "PLIN1", "PLIN4", "PLIN5", "CIDEA", "CIDEC", "LPL", "UCP1", "GHR",
          
          
          ## Endothelial cells
          
          # General endothelial cells
          "VWF", "PECAM1", "CDH5", "ANO2",
          
          # Vascular endothelial cells
          "VEGFC", "ROR2", "KDR",  "PLVAP", "EMCN",
          
          # Arterial endothelial cells
          "FLT1", "TEK", "EFNB2", "NOTCH4", "ANGPT2",
          
          # Lymphatic endothelial cells
          "PROX1", "FLT4",
          
          # Adhesion & cytoskeletal genes
          "EZR", "RBFOX1",
          
          
          ## Immune cells

          # Macrophages
          "MRC1", "F13A1", "CD163", "LYVE1", "SIGLEC1", "FRMD4B", "CD68", "ITGAM", "LPAR1", "SOX9",
          
          # T cells
          "BCL11B", "LCK", "PTPRC", "CD247", "CD3G",
          
          ## Muscle cells
          
          # vSMCs
          "PDGFRB", "NOTCH3", "GUCY1A2", "GUCY1A1", "MYH11", "MYO1D", "RYR2", "TRPC6", "ACTA2", "CNN1", "DES", 
          
          ## Other
          
          # Circadian rhythm
          "CRY2", "CSNK1E"
          
        )) +
  
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

# Save plot as PDF
if (!dir.exists("results/humanPVATsn")) dir.create("results/humanPVATsn")

pdf("results/humanPVATsn/cluster_dot_plot.pdf", width = 12, height = 8)
print(annot_dot_plot)
invisible(dev.off())

# Display plot
annot_dot_plot

```


## Cell type annotation 

First we plot the unannotated UMAP to visualise clusters. This might help in identifying cell types as more similar cell types tend to cluster together.

```{r fig.height=5, fig.width=10}

DefaultAssay(libs_combined_filt) <- "integrated"

unnannot_umap <- DimPlot(libs_combined_filt, reduction = "umap",  ncol = 3, label = TRUE) +
  ggtitle("Unannotated UMAP of PVAT Cells (snRNA-seq)") +
  theme(plot.title = element_text(face = "bold", size = 16))

# Save plot as PDF
if (!dir.exists("results/humanPVATsn/final_figs")) dir.create("results/humanPVATsn/final_figs")

pdf("results/humanPVATsn/final_figs/final_umap_nbs.pdf", width = 12, height = 8)
print(unnannot_umap)
invisible(dev.off())

# Display plot
unnannot_umap

```

We name the clusters with the corresponding cell type based on the dot plots generated above.

```{r}
anno_combined <- RenameIdents(libs_combined_filt, 
                              "0" = "Mature adipocytes", 
                              "1" = "Intermediate pre-adipocytes", 
                              "2" = "Vascular endothelial cells",
                              "3" = "Macrophages",
                              "4" = "Vascular endothelial cells",
                              "5" = "Vascular endothelial cells", 
                              "6" = "Early pre-adipocytes" , 
                              "7" = "Lymphatic endothelial cells", 
                              "8" = "PPARG+ EBF+ Vascular SMCs",
                              "9" = "Transitional adipocytes",  
                              "10" = "Vascular endothelial cells", 
                              "11" = "T cells",
                              "12" = "Vascular SMCs")
```

# Final plots

## Final UMAP plots

These are the final plots from the analysis.

### Unannotated UMAP plots

We show proper integration of all three samples by colouring the UMAP by sample origin.

```{r}

# Plot UMAP by sample type
umap_plot <- DimPlot(libs_combined_filt,
              reduction = "umap",
              group.by = "sampleType") +
  ggtitle("Unannotated UMAP Coloured by Sample (snRNA-seq)") +
  theme(plot.title = element_text(hjust = 0, face = "bold", size = 16))

# Save and display plot
if (!dir.exists("results/humanPVATsn/final_figs")) dir.create("results/humanPVATsn/final_figs")

pdf("results/humanPVATsn/final_figs/final_umap_integration.pdf", width = 12, height = 8)
print(umap_plot)
invisible(dev.off())

umap_plot

```


We plot the unannotated UMAP split by sample origin.

```{r fig.height=6, fig.width=14}

# Plot UMAP by sample type
plot1 <- DimPlot(libs_combined_filt,
              reduction = "umap",
              split.by = "sampleType",
              label = TRUE) +
  ggtitle("Unannotated UMAP of PVAT Cells Split By Sample (snRNA-seq)") +
  theme(plot.title = element_text(hjust = 0, face = "bold", size = 18))

# Save and display plot
if (!dir.exists("results/humanPVATsn/final_figs")) dir.create("results/humanPVATsn/final_figs")

pdf("results/humanPVATsn/final_figs/final_umap_by_sample.pdf", width = 14, height = 6)
print(plot1)
invisible(dev.off())

plot1

```

We plot the unannotated UMAP of all three samples integrated into one.

```{r fig.height=5, fig.width=10}

unnannot_umap <- DimPlot(libs_combined_filt, reduction = "umap", ncol = 3, label = TRUE) +
  ggtitle("Unannotated UMAP of PVAT Cells (snRNA-seq)") +
  theme(plot.title = element_text(face = "bold", size = 16))

# Save and display plot
if (!dir.exists("results/humanPVATsn/final_figs")) dir.create("results/humanPVATsn/final_figs")

pdf("results/humanPVATsn/final_figs/final_umap_nbs.pdf", width = 12, height = 8)
print(unnannot_umap)
invisible(dev.off())

unnannot_umap

```

### Annotated UMAP plots

We plot the annotated UMAP coloured by cluster identity.

```{r fig.height=5, fig.width=12}

# Define cluster colours
namedclusterColours <-
  list(
    "Mature adipocytes" = "#895129", 
    "Intermediate pre-adipocytes" = "#FFA500", 
    "Vascular endothelial cells" = "#56cafb",
    "Macrophages" = "#A300A3",
    "Early pre-adipocytes" = "#FFED29", 
    "Lymphatic endothelial cells" = "#00c04b", 
    "PPARG+ EBF+ Vascular SMCs" = "#FF1919",
    "Vascular SMCs" = "#FF6F61",
    "Transitional adipocytes" = "#BE6400",
    "T cells" = "#fdb0c0"
  )

# Set default assay
DefaultAssay(anno_combined) <- "integrated"

# Define custom order for legend
custom_order <- c("Early pre-adipocytes", "Intermediate pre-adipocytes", "Transitional adipocytes", "Mature adipocytes",
                  "PPARG+ EBF+ Vascular SMCs", "Vascular SMCs", "Vascular endothelial cells", "Lymphatic endothelial cells",
                  "Macrophages", "T cells")

# Ensure Idents are explicitly set with correct levels
Idents(anno_combined) <- factor(Idents(anno_combined), levels = custom_order)

# Generate UMAP plot with reordered legend
final_umap <- DimPlot(anno_combined,
                      cols = namedclusterColours) +
  ggtitle("UMAP Visualisation of Human PVAT Cell Populations (snRNA-seq)") +
  theme(plot.title = element_text(face = "bold", size = 16))
  

# Save and display plot
if (!dir.exists("results/humanPVATsn/final_figs")) dir.create("results/humanPVATsn/final_figs")

pdf("results/humanPVATsn/final_figs/final_umap_annot.pdf", width = 12, height = 8)
print(final_umap)
invisible(dev.off())

final_umap

```

## Final dot plot

We plot a dot plot showing the expression of marker genes across clusters.
- Dot size: percent expressed, i.e. how many cells in the given cluster express the gene.
- Dot colour: average expression, i.e. the expression level in cells that do express the given gene.

```{r fig.height = 7, fig.width = 16}

# Set the theme
t4 <- theme(                              
  plot.background = element_blank(), 
  panel.grid.major = element_line(linetype = "dotted", colour = "grey"),
  panel.grid.minor = element_blank(), 
  panel.border = element_blank(), 
  panel.background = element_blank(),
  axis.line = element_line(linewidth = .4),
  axis.line.x = element_line(),
  axis.line.y = element_line(),
  axis.text.x  = element_text(angle = 90, size = 15, hjust = 1, vjust = 0.5),
  axis.text.y = element_text(size = 15),
  axis.ticks.x = element_line(linetype = "dotted"),
  axis.ticks.length = unit(0.3,"cm"),
  axis.title.x = element_text(face="bold", size=18), 
  axis.title.y = element_text(face="bold", size=18),
  plot.title = element_text(face="bold", size=20),
  strip.text = element_text(size = 10),
  legend.title = element_text(size = 18),
  legend.text = element_text(size = 15)
)

# Set factor levels of cellular identities
anno_combined2 <- anno_combined
Idents(anno_combined2) <- factor(Idents(anno_combined2),
                                 levels = c("T cells", "Macrophages", "Lymphatic endothelial cells", "Vascular endothelial cells", 
                                            "Vascular SMCs", "PPARG+ EBF+ Vascular SMCs", "Mature adipocytes", "Transitional adipocytes",
                                            "Intermediate pre-adipocytes", "Early pre-adipocytes"))

# Set default assay
DefaultAssay(anno_combined2) <- "RNA"

# Define marker genes to plot
markers_to_plot <- c(
          ## Adipocyte-related
          
          # Pre-adipocytes
          "NEGR1", "EBF2", "ABCA10", "DCN", "MGP", "PDGFRA", "FBLN1", "COL1A1", "COL1A2", "FSTL1", "CREB5", "PPARG",
          # Mature brown adipocytes
          "ADIPOQ", "PLIN1", "PLIN4", "PLIN5", "CIDEA", "CIDEC", "LPL", "GHR",
          
          ## Endothelial cells
          # General endothelial cells
          "VWF", "PECAM1", "CDH5", "ANO2",
          # Vascular endothelial cells
          "VEGFC", "ROR2", "KDR",  "PLVAP", "EMCN",
          # Arterial endothelial cells
          "FLT1", "TEK", "EFNB2", "NOTCH4",
          # Lymphatic endothelial cells
          "PROX1", "FLT4",
          
          ## Immune cells
          # Macrophages
          "MRC1", "F13A1", "CD163", "LYVE1", "SIGLEC1", "FRMD4B",
          # T cells
          "BCL11B", "LCK", "PTPRC", "CD247",
          
          ## Muscle cells
          # vSMCs
          "PDGFRB", "NOTCH3", "GUCY1A2", "GUCY1A1", "MYH11", "MYO1D", "RYR2", "TRPC6", "ACTA2", "CNN1", 
          
          ## Other
          # Circadian rhythm
          "CRY2", "CSNK1E"
        )

# Generate dot plot
final_dotplot <- DotPlot(anno_combined2, features = markers_to_plot, dot.scale = 7, cols = "PiYG") +
  ggtitle("Expression of Marker Genes Across PVAT Cell Types (snRNA-seq)") + 
  t4 + 
  xlab("Marker Gene") + ylab("Cell Type")

# Save and display plot
if (!dir.exists("results/humanPVATsn/final_figs")) dir.create("results/humanPVATsn/final_figs")

pdf("results/humanPVATsn/final_figs/final_marker_dot_plot.pdf", width = 17, height = 8)
print(final_dotplot)
invisible(dev.off())

final_dotplot

```


## Final feature plots

We plot feature plots of marker genes to show the expression specific to some clusters.

```{r fig.height=15, fig.width=15}

DefaultAssay(anno_combined) <- "RNA"

feature_plot <- FeaturePlot(anno_combined,
                            features = c(
                              # Pre-adipocyte
                              "ABCA10","PDGFRA", "EBF2", 
                              # Mature adipocyte
                              "PLIN4", "CIDEA", "ADIPOQ",
                              # vSMC
                              "MYH11",
                              # Endothelial
                              "PECAM1",
                              # Macrophages
                              "MRC1"
                              ),
                            min.cutoff = 1)

final_feature <- feature_plot + plot_annotation(title = "Feature Plot of Marker Genes of PVAT Cell Populations (snRNA-seq)",
                  theme = theme(plot.title = element_text(face="bold", size=20)))

# Save the figure
if (!dir.exists("results/humanPVATsn/final_figs")) dir.create("results/humanPVATsn/final_figs")

pdf("results/humanPVATsn/final_figs/feature_plots.pdf", width = 15, height = 15)
print(final_feature)
invisible(dev.off())

# Display the plot
final_feature

```


## Final violin plots

We plot violin plots of marker genes to show distribution of expression across clusters.

```{r fig.width=30, fig.height=20}

# Set default assay to RNA for expression plots
DefaultAssay(anno_combined) <- "RNA"

# Define theme
t1 <- theme(
  plot.title = element_text(size = 22, face = "bold", hjust = 0.5),
  legend.position = "right",
  legend.text = element_text(size = 20),
  plot.margin = margin(10, 10, 10, 50)
)

# Define genes to plot
features_to_plot <- c("EBF2", "PPARG", "PDGFRA", "ADIPOQ", "MYH11", "PDGFRB")

# Generate violin plots
violin_plots <- lapply(features_to_plot, function(gene) {
  VlnPlot(anno_combined, features = gene, pt.size = 0, cols = unlist(namedclusterColours)) +
    theme(
      axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 14),
      axis.text.y = element_text(size = 14),
      axis.title.x = element_blank(),
      axis.title.y = element_text(size = 16, face = "bold"),
      plot.title = element_text(size = 16, face = "bold", hjust = 0.5)
    )
})

# Arrange plots in a 2x3 grid and add main title
final_violin_plot <- wrap_plots(violin_plots, ncol = 3, nrow = 2, guides = "collect") +
  plot_annotation(title = "Violin Plot of Marker Gene Expression in PVAT",
                  theme = t1)

# Save figure
if (!dir.exists("results/humanPVATsn/final_figs")) dir.create("results/humanPVATsn/final_figs")

pdf("results/humanPVATsn/final_figs/violin_plots.pdf", width = 25, height = 15)
print(final_violin_plot)
invisible(dev.off())

# Display plot
final_violin_plot

```

# Save final integrated Seurat object

```{r}

save(anno_combined, file = "data/analysis/anno_combined.RData")

```


# Session info

**Session Info**

```{r}

sessionInfo()

```
