---
title: "Pathway enrichment analysis of human perivascular adipose tissue (PVAT) snRNA-seq data"
author: "Hannah M. L. Bazin"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
---

In this vignette, I run `pathfindR` on differentially expressed genes of single-nucleus RNA sequencing of human perivascular adipose tissue (PVAT). The aim is to identify enriched pathways in each stage of brown adipocyte development. The raw data is available on the Gene Expression Omnibus website under the accession number GSE164528, found [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE164528).


# Overview of `pathfindR` algorithm

`pathfindR` is an R package designed for active-subnetwork-oriented pathway enrichment analysis. Unlike traditional over-representation analysis (ORA) and gene set enrichment analysis (GSEA), which evaluate gene lists without considering gene interactions, pathfindR integrates protein-protein interaction networks (PINs) to identify active subnetworks – groups of interacting genes enriched in significantly altered genes. A subnetwork is defined as a cluster of interconnected genes in a PIN, and it is considered active if it contains a disproportionately high number of differentially expressed genes. The algorithm then performs pathway enrichment analysis on these subnetworks to identify biologically relevant pathways.

The `pathfindR` workflow consists of the following steps:

1.  **Mapping input genes** and their p-values onto a predefined PIN.
2.  **Identification of active subnetworks**, using a heuristic search algorithm to detect interconnected gene clusters enriched in significant genes.
3.  **Filtering subnetworks** based on predefined scoring criteria, including (i) the number of significant genes (minimum of 3 by default) and (ii) the subnetwork score, calculated as the sum of absolute log-transformed p-values of significant genes.
4.  **Pathway enrichment analysis** on selected subnetworks using Fisher’s Exact Test, with multiple testing correction using the Bonferroni method by default.
5.  **Iterative analysis**, where the active subnetwork search and enrichment steps are repeated multiple times to account for random variation in subnetwork detection, ensuring robustness by identifying pathways that remain significantly enriched across multiple runs.
6.  **Output generation**, producing a structured data frame of significantly enriched pathways and gene sets, along with key metrics such as the lowest and highest adjusted p-values across iterations, to assess the reproducibility of pathway enrichment across iterations.

# Set up

## Set working directory

The working directory can be set by running the following command in the console: `setwd("/Users/hannahbazin/Desktop/Cambridge/Academics/Han_Lab/MPhil/mphil-project")`

## Load libraries

```{r}

library(Seurat)
library(pathfindR)
library(tidyverse)
library(enrichplot)
library(dplyr)
library(ggtext)
library(patchwork)

```

## Set seed

Because `pathfindR` includes a heuristic search algorithm to identify active subnetworks, we set a random seed for reproducibility.

```{r}

set.seed(1996)

```


## Load data

We reload the integrated Seurat object generated in the notebook titled `human_PVAT_snRNAseq.Rmd`.

```{r}

# Load integrated Seurat object from previous analysis
load(file = "data/analysis/anno_combined.RData")

```

# Pre-processing workflow

## Subset adipocyte clusters

Since the aim is to identify differentially activated pathways in the four adipocyte-related clusters, we subset the Seurat object to contain only these clusters.

```{r}

# Select clusters in adipocyte lineage
adipo_clusters <- c("Early pre-adipocytes",
                    "Intermediate pre-adipocytes",
                    "Transitional adipocytes",
                    "Mature adipocytes")

# Subset Seurat object to include only adipocyte clusters
anno_combined_adipo <- subset(anno_combined, idents = adipo_clusters)

```

## Scale data

Subsetting a Seurat object can cause the loss of the scaled data in the object. We verify this and recompute the scaled data as needed. This is a time-consuming step, so it is best to save the R object.

```{r}

# Ensure default assay is RNA assay
DefaultAssay(anno_combined_adipo) <- "RNA"

# Check that scaled data was preserved - this can be lost while subsetting
head(anno_combined_adipo@assays$RNA@scale.data) # <0 x 0 matrix>

# Scale data if needed
  # This is a time-consuming step, save the object
if (file.exists("data/analysis/anno_combined_adipo_scaled.RData")) {
  
  message("Loading existing scaled data file.")
  
  load(file = "data/analysis/anno_combined_adipo_scaled.RData")
  
} else {
  
  message("Missing scaled data file. Scaling data now.")
  
  anno_combined_adipo_scaled <- ScaleData(anno_combined_adipo, vars.to.regress = "percent.mt")
  
  save(anno_combined_adipo_scaled, file = "data/analysis/anno_combined_adipo_scaled.RData")

}

```

# Find markers

The Seurat `FindMarkers` function identifies differentially expressed genes by comparing one cluster with another. The output of this function is a dataframe with markers as rows with associated statistics. The columns are:
- `avg_logFC`: log fold-change of the average expression between the groups. If positive, the gene is more highly expressed in the first group.
- `pct.1`: percentage of cells expressing the gene in the first group
- `pct.2`: percentage of cells expressing the gene in the second group
- `p_val_adj`: adjusted p-value after Bonferroni correction

To run the full analysis from scratch, be sure to delete any existing marker files in the folder `results/humanPVATsn/pathfindR/full/`.

### Set up

```{r}

# Ensure default assay is RNA assay
DefaultAssay(anno_combined_adipo_scaled) <- "RNA"

marker_files <- c("results/humanPVATsn/pathfindR/full/early_vs_inter_markers.csv",
                  "results/humanPVATsn/pathfindR/full/inter_vs_trans_markers.csv",
                  "results/humanPVATsn/pathfindR/full/trans_vs_mature_markers.csv",
                  "results/humanPVATsn/pathfindR/full/early_vs_mature_markers.csv",
                  "results/humanPVATsn/pathfindR/full/early_vs_mature_markers_inv.csv"
                  )

```

The `FindMarkers` function is run on each stage of brown adipocyte development, including on the overall development (early vs mature).

### Early vs intermediate pre-adipocytes

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[1])) {
  
  early_vs_inter_markers <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Early pre-adipocytes",
                                      ident.2 = "Intermediate pre-adipocytes",
                                      min.pct = 0.3,
                                      logfc.threshold = 0.3
                                      )
  
  write.csv(early_vs_inter_markers, marker_files[1], row.names = TRUE)
  
  message(paste0("Saved markers for early vs intermediate pre-adipocytes to ", marker_files[1]))
  
} else {
  
  message("Marker file for early vs intermediate pre-adipocytes already exists. Skipping FindMarkers.")
  
}

```

### Intermediate pre-adipocytes vs transitional adipocytes

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[2])) {
  
  inter_vs_trans_markers <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Intermediate pre-adipocytes",
                                      ident.2 = "Transitional adipocytes",
                                      min.pct = 0.3,
                                      logfc.threshold = 0.3
                                      )
  
  write.csv(inter_vs_trans_markers, marker_files[2], row.names = TRUE)
  
  message(paste0("Saved markers for intermediate pre-adipocytes vs transitional adipocytes to ", marker_files[2]))
  
} else {
  
  message("Marker file for intermediate pre-adipocytes vs transitional adipocytes already exists. Skipping FindMarkers.")
  
}

```

### Transitional adipocytes vs mature adipocytes

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[3])) {
  
  trans_vs_mature_markers <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Transitional adipocytes",
                                      ident.2 = "Mature adipocytes",
                                      min.pct = 0.3,
                                      logfc.threshold = 0.3
                                      )
  
  write.csv(trans_vs_mature_markers, marker_files[3], row.names = TRUE)
  
  message(paste0("Saved markers for transitional vs mature adipocytes to ", marker_files[3]))
  
} else {
  
  message("Marker file for transitional vs mature adipocytes already exists. Skipping FindMarkers.")
  
}

```

### Early pre-adipocytes vs mature adipocytes

This first analysis of the full differentiation process looks at what genes are upregulated in the early pre-adipocytes versus the mature adipocytes.

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[4])) {
  
  early_vs_mature_markers <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Early pre-adipocytes",
                                      ident.2 = "Mature adipocytes",
                                      min.pct = 0.3,
                                      logfc.threshold = 0.3
                                      )
  
  write.csv(early_vs_mature_markers, marker_files[4], row.names = TRUE)
  
  message(paste0("Saved markers for early pre-adipocytes vs mature adipocytes to ", marker_files[4]))
  
} else {
  
  message("Marker file for early pre-adipocytes vs mature adipocytes already exists. Skipping FindMarkers.")
  
}

```

### Mature adipocytes vs early pre-adipocytes

This second analysis of the full differentiation process looks at what genes are upregulated in the mature adipocytes versus the early pre-adipocytes.

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[5])) {
  
  early_vs_mature_markers_inv <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Mature adipocytes",
                                      ident.2 = "Early pre-adipocytes",
                                      min.pct = 0.3,
                                      logfc.threshold = 0.3
                                      )
  
  write.csv(early_vs_mature_markers_inv, marker_files[5], row.names = TRUE)
  
  message(paste0("Saved markers for mature adipocytes vs early pre-adipocytes to ", marker_files[5]))
  
} else {
  
  message("Marker file for mature adipocytes vs early pre-adipocytes already exists. Skipping FindMarkers.")
  
}

```

# Pathway enrichment analysis

## Load marker files

To streamline the workflow the marker files are saved as `csv` files in the folder `results/humanPVATsn/pathfindR/full/`. These can be loaded into RStudio to save the user from running FindMarkers() again, which is a time-consuming step.

```{r}

# Redefine marker files if starting code from this chunk
marker_files <- c("results/humanPVATsn/pathfindR/full/early_vs_inter_markers.csv",
                  "results/humanPVATsn/pathfindR/full/inter_vs_trans_markers.csv",
                  "results/humanPVATsn/pathfindR/full/trans_vs_mature_markers.csv",
                  "results/humanPVATsn/pathfindR/full/early_vs_mature_markers.csv",
                  "results/humanPVATsn/pathfindR/full/early_vs_mature_markers_inv.csv"
                  )

# Load all relevant marker files
if (all(file.exists(marker_files))) {
  
  early_vs_inter_markers <- read_csv(marker_files[1])
  inter_vs_trans_markers <- read_csv(marker_files[2])
  trans_vs_mature_markers <- read_csv(marker_files[3])
  early_vs_mature_markers <- read_csv(marker_files[4])
  early_vs_mature_markers_inv <- read_csv(marker_files[5])
  
} else {
  
  stop("One or more marker files are missing. Run FindMarkers first.")
  
}

```

## Reformat data

The data needs to be reformatted to align with the expected input of `pathfindR`. The input data frame must consist of columns containing: gene symbols, change values (optional) and p values.

```{r}

# Make function to reformat to pathfindR input
reformat_pathfindR <- function(input) {
  
  input %>%
    select(Gene.symbol = ...1, logFC = avg_log2FC, adj.P.Val = p_val_adj) %>% 
    as.data.frame()

}

# Reformat markers
early_vs_inter_markers <- reformat_pathfindR(early_vs_inter_markers)
inter_vs_trans_markers <- reformat_pathfindR(inter_vs_trans_markers)
trans_vs_mature_markers <- reformat_pathfindR(trans_vs_mature_markers)
early_vs_mature_markers <- reformat_pathfindR(early_vs_mature_markers)
early_vs_mature_markers_inv <- reformat_pathfindR(early_vs_mature_markers_inv)

```

## Run `pathfindR`

For early vs intermediate pre-adipocytes.

```{r}

step1_output <- run_pathfindR(early_vs_inter_markers,
                              gene_sets = "GO-BP",
                              min_gset_size = 5,
                              max_gset_size = 300
                              )

```

For intermediate pre-adipocytes vs transitional adipocytes.

```{r}

step2_output <- run_pathfindR(inter_vs_trans_markers,
                              gene_sets = "GO-BP",
                              min_gset_size = 5,
                              max_gset_size = 300
                              )

```

For transitional vs mature adipocytes.

```{r}

step3_output <- run_pathfindR(trans_vs_mature_markers,
                              gene_sets = "GO-BP",
                              min_gset_size = 5,
                              max_gset_size = 300
                              )

```

For early pre-adipocytes vs mature adipocytes.

In this output dataframe, a positive fold enrichment score indicates a pathway enriched in the early pre-adipocytes compared to the mature adipocytes.

```{r}

full_diff_output <- run_pathfindR(early_vs_mature_markers,
                                  gene_sets = "GO-BP",
                                  min_gset_size = 5,
                                  max_gset_size = 300
                                  )

```

In this output dataframe, a positive fold enrichment score indicates a pathway enriched in the mature adipocytes compared to the early pre-adipocytes.

```{r}

full_diff_output_inv <- run_pathfindR(early_vs_mature_markers_inv,
                                  gene_sets = "GO-BP",
                                  min_gset_size = 5,
                                  max_gset_size = 300
                                  )

```

## Compare `pathfindR` results

The `pathfindR` package provides a function to compare two different `pathfindR` output dataframes. More details can be found in [this vignette](https://cran.r-project.org/web/packages/pathfindR/vignettes/comparing_results.html).

We compare pathway activation between the first and second step of adipocyte differentiation: early pre-adipocytes to intermediate pre-adipocytes and intermediate pre-adipocytes to transitional adipocytes.

```{r}

compare_steps1v2 <- combine_pathfindR_results(
  result_A = step1_output,
  result_B = step2_output,
  plot_common = FALSE
)

```

We compare pathway activation between the second and third (final) step of adipocyte differentiation: intermediate pre-adipocytes to transitional adipocytes and transitional adipocytes to mature adipocytes.

```{r}

compare_steps2v3 <- combine_pathfindR_results(
  result_A = step2_output,
  result_B = step3_output,
  plot_common = FALSE
)

```

## Save files

We save the files to streamline the workflow, as this is a time-consuming process.

```{r}

# Step-specific enriched pathways - for running cluster_enriched_terms()
saveRDS(step1_output, "results/humanPVATsn/pathfindR/full/step1.rds")
saveRDS(step2_output, "results/humanPVATsn/pathfindR/full/step2.rds")
saveRDS(step3_output, "results/humanPVATsn/pathfindR/full/step3.rds")
saveRDS(full_diff_output, "results/humanPVATsn/pathfindR/full/full_diff.rds")
saveRDS(full_diff_output_inv, "results/humanPVATsn/pathfindR/full/full_diff_inv.rds")

# Comparison dataframes - for visualisation in bar plots
write.csv(compare_steps1v2, "results/humanPVATsn/pathfindR/full/step1_vs_step2.csv", row.names = TRUE)
write.csv(compare_steps2v3, "results/humanPVATsn/pathfindR/full/step2_vs_step3.csv", row.names = TRUE)
write.csv(full_diff_output, "results/humanPVATsn/pathfindR/full/full_diff.csv", row.names = TRUE)

```

# Visualisation

Bar plots are generated to visualise the top 20 enriched pathways for each of the comparisons.

## Load data

```{r}

# Load data
step1_vs_step2 <- read_csv("results/humanPVATsn/pathfindR/full/step1_vs_step2.csv")
step2_vs_step3 <- read_csv("results/humanPVATsn/pathfindR/full/step2_vs_step3.csv")
full_diff <- readRDS("results/humanPVATsn/pathfindR/full/full_diff.rds")

```

## Create bar plots

### Define function

Define function to create and save bar plots.
- For the A only and B only pathways, enriched terms are ranked based on fold enrichment.
- For the common pathways, enriched terms are ranked based on the combined p-value; this helps identify pathways that are consistently significant in both datasets.


```{r}

create_and_save_bar_plot <- function(df, category, title_label, output_file, enriched_step) {

  # Assign colour based on step
  fill_col <- case_when(
    enriched_step == "step1" ~ "#FDD866",
    enriched_step == "step2" ~ "#F46D43",
    enriched_step == "step3" ~ "#D73027",
    enriched_step == "common" ~ "#CC79A7"
  )

  # Differentiate cases for common or A only and B only enriched pathways
  if(category == "common") {
    
    sorting_col <- "combined_p"
    
    # Filter top 20 enriched pathways
    df_filtered <- df %>% 
      filter(status == category) %>%
      # Sort by descending combined p value
      arrange(desc(combined_p)) %>%
      # Keep only those with more than 5 occurrences
      filter(occurrence_A > 5 & occurrence_B > 5) %>%
      # Keep top 20 enriched pathways
      slice_head(n = 20) %>%
      # Ensure factor levels are correctly ordered for plotting
      mutate(Term_Description = factor(Term_Description, levels = Term_Description))
    
    # Adjust x-axis for "common" category to ensure longest bars at the top
    x_var <- -log10(df_filtered[[sorting_col]])  # Convert p-values for better visualisation
    x_label <- "-log10(combined P value)"
    
  } else {
    
    sorting_col <- paste0("Fold_Enrichment_", substr(category, 1, 1))
    occurrence_col <- paste0("occurrence_", substr(category, 1, 1))
    
    # Filter top 20 enriched pathways
    df_filtered <- df %>%
      filter(status == category) %>%
      # Sort by descending fold enrichment
      arrange(desc(.data[[sorting_col]])) %>%
      # Keep only those with more than 5 occurrences
      filter(.data[[occurrence_col]] > 5) %>% 
      # Keep top 20 enriched pathways
      slice_head(n = 20) %>%
      # Ensure factor levels are correctly ordered for plotting
      mutate(Term_Description = factor(Term_Description, levels = rev(Term_Description)))
        
    x_var <- df_filtered[[sorting_col]]
    x_label <- "Fold enrichment"
    
  }
  
  # Create bar plot
  plot <- ggplot(df_filtered, aes(x = x_var, y = Term_Description)) +
    geom_bar(stat = "identity", color = "black", fill = fill_col) +
    labs(title = title_label, x = x_label, y = "Pathway") +
    theme_minimal() +
    theme(legend.position = "none", plot.margin = margin(10, 20, 10, 10))
  
  # Save plot
  ggsave(output_file, plot, width = 13, height = 6, device = "pdf")

  # Return plot
  return(plot)
  
}

# Define function to obtain enriched step for bar plot colour
get_enriched_step <- function(category, comparison) {
  if (category == "common") {
    return("common")
  } else if (comparison == "step1_vs_step2") {
    return(ifelse(condition == "A only", "step1", "step2"))
  } else if (comparison == "step2_vs_step3") {
    return(ifelse(condition == "A only", "step2", "step3"))
  }
}

```

### Call function

Call bar plot function for each comparison.

```{r fig.width=14, fig.height=6}

# List of conditions
conditions <- c("A only", "B only", "common")

# Condition labels for each comparison
condition_labels1 <- list("A only" = "step 1 only", "B only" = "step 2 only", "common" = "both steps")
condition_labels2 <- list("A only" = "step 2 only", "B only" = "step 3 only", "common" = "both steps")

# Create a bar plot for each condition defined above
for (condition in conditions) {
  
  # Define enriched step for bar plot colour
  enriched_step1 <- get_enriched_step(condition, "step1_vs_step2")
  enriched_step2 <- get_enriched_step(condition, "step2_vs_step3")
  
  # Bar plot for comparing steps 1 and 2 of differentiation
  step1_vs_step2_plot <- create_and_save_bar_plot(step1_vs_step2, condition,
                                                  paste("Step 1 vs 2 of brown adipocyte differentiation: top 20 enriched pathways for", condition_labels1[[condition]]),
                                                  paste0("results/humanPVATsn/pathfindR/full/step1_vs_step2_", gsub(" ", "_", condition_labels1[[condition]]), ".pdf"),
                                                  enriched_step1)
  
  print(step1_vs_step2_plot)
  
  # Bar plot for comparing steps 2 and 3 of differentiation
  step2_vs_step3_plot <- create_and_save_bar_plot(step2_vs_step3, condition,
                                                    paste("Step 2 vs 3 of brown adipocyte differentiation: top 20 enriched pathways for", condition_labels2[[condition]]),
                                                  paste0("results/humanPVATsn/pathfindR/full/step2_vs_step3_", gsub(" ", "_", condition_labels2[[condition]]), ".pdf"),
                                                  enriched_step2)
  
  print(step2_vs_step3_plot)
  
}

```

Visualise results for differentially activated pathways for the full differentiation process.

This shows the enriched pathways in the early pre-adipocytes compared to the mature adipocytes.

```{r fig.width=12, fig.height=6}

full_diff_filtered <- full_diff %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

plot <- ggplot(full_diff_filtered,
                 aes(x = .data[["Fold_Enrichment"]], y = Term_Description)) +
    geom_bar(stat = "identity", color = "black", fill = "#2166AC") +
    labs(title = "Full differentiation: top 20 enriched pathways",
         x = "Fold enrichment in early pre-adipocytes",
         y = "Pathway") +
    theme_minimal() +
    theme(legend.position = "none",
          plot.margin = margin(10, 20, 10, 10))
  
# Save plot
ggsave("results/humanPVATsn/pathfindR/full/full_diff.pdf", plot, width = 12, height = 6, device = "pdf")

# Show plot
print(plot)

```

### Colour bars by theme

After manually sorting the pathways into biological themes, the plots are coloured by theme to improve interpretability. This was to see how it looked, but I chose to visualise the plots in the next way, by facet wrapping them - see code block below.


```{r fig.width=12, fig.height=6}

full_diff_filtered <- full_diff %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

# Biological themes
full_diff_filtered$Theme <- c(
  "Signal transduction & cell communication",  # 1
  "Metabolic activation",                      # 2
  "Metabolic activation",                      # 3
  "Cellular restructuring & vesicle dynamics", # 4
  "Metabolic activation",                      # 5
  "Cellular restructuring & vesicle dynamics", # 6
  "Metabolic activation",                      # 7
  "Differentiation & tissue remodelling",      # 8
  "Differentiation & tissue remodelling",      # 9
  "Cellular restructuring & vesicle dynamics", # 10
  "Cellular restructuring & vesicle dynamics", # 11
  "Cellular restructuring & vesicle dynamics", # 12
  "Metabolic activation",                      # 13
  "Signal transduction & cell communication",  # 14
  "Signal transduction & cell communication",  # 15
  "Signal transduction & cell communication",  # 16
  "Differentiation & tissue remodelling",      # 17
  "Signal transduction & cell communication",  # 18
  "Metabolic activation",                      # 19
  "Differentiation & tissue remodelling"       # 20
)

# Count number of pathways per theme
theme_counts <- full_diff_filtered %>%
  count(Theme) %>%
  arrange(desc(n))

# Reorder Theme levels based on size (most pathways first)
full_diff_filtered$Theme <- factor(full_diff_filtered$Theme,
                                   levels = theme_counts$Theme)

# Sort data frame first by Theme (ordered by count), then by Term_Description
full_diff_filtered <- full_diff_filtered %>%
  arrange(Theme, Term_Description) %>%
  mutate(Term_Description = factor(Term_Description, levels = Term_Description))

# Plot coloured by biological theme
plot <- ggplot(full_diff_filtered,
               aes(x = Fold_Enrichment, y = Term_Description, fill = Theme)) +
  geom_bar(stat = "identity", color = "black") +
  labs(title = "Full differentiation: top 20 enriched pathways",
       x = "Fold enrichment in early preadipocytes",
       y = "Pathway",
       fill = "Biological theme") +  # add legend title
  theme_minimal() +
  theme(plot.margin = margin(10, 20, 10, 10))

# Optional: Uncomment to use predefined color palette (e.g., "Set3")
# plot <- plot + scale_fill_brewer(palette = "Set3")

# Optional: Uncomment to use custom colors
# plot <- plot + scale_fill_manual(values = c("Theme1" = "#E41A1C", "Theme2" = "#377EB8", "Theme3" = "#4DAF4A"))

# Save and show plot
ggsave("results/humanPVATsn/pathfindR/full/full_diff_colored.pdf", plot, width = 12, height = 6)
print(plot)


```

### Split pathways by theme

Step 1

```{r fig.width=12, fig.height=10}

# Filter pathfindR output
step1_filtered <- step1_output %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only significantly enriched pathways
  filter(lowest_p < 0.05) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

# Biological theme list
theme_list <- c(
  "Cellular & tissue \n remodelling",
  "Cell signalling",
  "Neurogenic priming",
  "Metabolic & transport \n processes"
)

# Assign themes
step1_filtered$Theme <- c(
  theme_list[2],
  theme_list[3],
  theme_list[3],
  theme_list[1],
  theme_list[3],
  theme_list[2],
  theme_list[4],
  theme_list[1],
  theme_list[1],
  theme_list[2],
  theme_list[1],
  theme_list[1],
  theme_list[1],
  theme_list[4],
  theme_list[2],
  theme_list[1],
  theme_list[4],
  theme_list[4],
  theme_list[2],
  theme_list[1]
)

# Set Theme factor levels to control facet order
step1_filtered$Theme <- factor(step1_filtered$Theme, levels = theme_list)

# Reorder by theme and pathway name
step1_filtered <- step1_filtered %>%
  arrange(Theme, Term_Description) %>%
  mutate(Term_Description = factor(Term_Description, levels = Term_Description))

# Create facetted bar plot
plot_base <- ggplot(step1_filtered,
               aes(x = Fold_Enrichment, y = Term_Description)) +
  geom_col(fill = "#FDD866", color = "black") +
  facet_wrap(~ Theme, scales = "free_y", ncol = 1, strip.position = "left") +
  labs(x = "Fold enrichment", y = NULL) +
  theme_minimal(base_size = 12) +
  coord_cartesian(clip = "off") + 
  theme(
    strip.text = element_text(size = 12, face = "bold"),
    strip.placement = "outside",
    strip.background = element_blank(),
    panel.spacing = unit(1.5, "lines"),
    axis.text.y = element_text(size = 10),
    plot.margin = margin(10, 50, 10, 50),
    plot.title = element_text(hjust = 0.5)
  )

# Use patchwork for plot title
plot <- plot_base + 
  plot_annotation(
    title = "Top pathways enriched in step 1 (early to intermediate preadipocytes)",
    theme = theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))
  )

# Save and show plot
ggsave("results/humanPVATsn/pathfindR/full/step1_split.pdf", plot, width = 12, height = 9)
print(plot)


```

Step 2

```{r fig.width=12, fig.height=10}

# Filter pathfindR output
step2_filtered <- step2_output %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only significantly enriched pathways
  filter(lowest_p < 0.05) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

# Biological theme list
theme_list <- c(
  "Metabolic \n priming",
  "Cell signalling & \n tissue remodelling",
  "Neurogenic \n priming",
  "Developmental \n origins",
  "Amyloid-beta \n processing"
)


# Assign themes
step2_filtered$Theme <- c(
  theme_list[1],
  theme_list[1],
  theme_list[4],
  theme_list[1],
  theme_list[2],
  theme_list[1],
  theme_list[2],
  theme_list[2],
  theme_list[4],
  theme_list[1],
  theme_list[1],
  theme_list[1],
  theme_list[1],
  theme_list[5],
  theme_list[1],
  theme_list[2],
  theme_list[3],
  theme_list[3],
  theme_list[3],
  theme_list[1]
)

# Set Theme factor levels to control facet order
step2_filtered$Theme <- factor(step2_filtered$Theme, levels = theme_list)

# Reorder by theme and pathway name
step2_filtered <- step2_filtered %>%
  arrange(Theme, Term_Description) %>%
  mutate(Term_Description = factor(Term_Description, levels = Term_Description))

# Create facetted bar plot
plot_base <- ggplot(step2_filtered,
               aes(x = Fold_Enrichment, y = Term_Description)) +
  geom_col(fill = "#F46D43", color = "black") +
  facet_wrap(~ Theme, scales = "free_y", ncol = 1, strip.position = "left") +
  labs(x = "Fold enrichment", y = NULL) +
  theme_minimal(base_size = 12) +
  coord_cartesian(clip = "off") + 
  theme(
    strip.text = element_text(size = 12, face = "bold"),
    strip.placement = "outside",
    strip.background = element_blank(),
    panel.spacing = unit(1.5, "lines"),
    axis.text.y = element_text(size = 10),
    plot.margin = margin(10, 50, 10, 50),
    plot.title = element_text(hjust = 0.5)
  )

# Use patchwork for plot title
plot <- plot_base + 
  plot_annotation(
    title = "Top pathways enriched in step 2 (intermediate preadipocytes to transitional adipocytes)",
    theme = theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))
  )

# Save and show plot
ggsave("results/humanPVATsn/pathfindR/full/step2_split.pdf", plot, width = 12, height = 9)
print(plot)

```


Full differentiation

```{r fig.width=12, fig.height=9}

full_diff_filtered <- full_diff %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

# Biological theme list
theme_list <- c(
  "Metabolic activation",
  "Cellular restructuring \n & vesicle dynamics",
  "Signal transduction \n & cell communication",
  "Differentiation \n & tissue remodelling"
)

# Assign themes using numeric indexing
full_diff_filtered$Theme <- c(
  theme_list[3],
  theme_list[1],
  theme_list[1],
  theme_list[2],
  theme_list[1],
  theme_list[2],
  theme_list[1],
  theme_list[4],
  theme_list[4],
  theme_list[2],
  theme_list[2],
  theme_list[2],
  theme_list[1],
  theme_list[3],
  theme_list[3],
  theme_list[3],
  theme_list[4],
  theme_list[3],
  theme_list[1],
  theme_list[4]
)

# Set Theme factor levels to control facet order
full_diff_filtered$Theme <- factor(full_diff_filtered$Theme, levels = theme_list)

# Reorder by theme and pathway name
full_diff_filtered <- full_diff_filtered %>%
  arrange(Theme, Term_Description) %>%
  mutate(Term_Description = factor(Term_Description, levels = Term_Description))

# Create facetted bar plot
plot1 <- ggplot(full_diff_filtered,
               aes(x = Fold_Enrichment, y = Term_Description)) +
  geom_col(fill = "#2166AC", color = "black") +
  facet_wrap(~ Theme, scales = "free_y", ncol = 1, strip.position = "left") +
  labs(x = "Fold enrichment", y = NULL) +
  theme_minimal(base_size = 12) +
  coord_cartesian(clip = "off") + 
  theme(
    strip.text = element_text(size = 12, face = "bold"),
    strip.placement = "outside",
    strip.background = element_blank(),
    panel.spacing = unit(1.5, "lines"),
    axis.text.y = element_text(size = 10),
    plot.margin = margin(10, 50, 10, 50),
    plot.title = element_text(hjust = 0.5)
  )

# Use patchwork for plot title
plot <- plot1 + 
  plot_annotation(
    title = "Top pathways enriched from early to mature brown adipocytes",
    theme = theme(plot.title = element_text(hjust = 0.5, size = 14, face = "bold"))
  )

# Save and show plot
ggsave("results/humanPVATsn/pathfindR/full/full_diff_split.pdf", plot, width = 12, height = 9)
print(plot)

```


# Cluster enriched terms

The function `cluster_enriched_terms` from the `pathfindR` package clusters enriched pathways into biologically relevant umbrella terms. This is helpful for the downstream biological interpretation. It uses pairwise kappa statistics to perform hierarchical clustering of enriched terms to determine the representative term.

## Load data

We load the previously saved RDS objects.

```{r}

# Load data
step1_output <- readRDS(file = "results/humanPVATsn/pathfindR/full/step1.rds")
step2_output <- readRDS(file = "results/humanPVATsn/pathfindR/full/step2.rds")
step3_output <- readRDS(file = "results/humanPVATsn/pathfindR/full/step3.rds")
full_diff_output <- readRDS(file = "results/humanPVATsn/pathfindR/full/full_diff.rds")

```


## visualisation from pathfindR

Reload marker files

```{r}

# Redefine marker files if starting code from this chunk
marker_files <- c("results/humanPVATsn/pathfindR/full/early_vs_inter_markers.csv",
                  "results/humanPVATsn/pathfindR/full/inter_vs_trans_markers.csv",
                  "results/humanPVATsn/pathfindR/full/trans_vs_mature_markers.csv",
                  "results/humanPVATsn/pathfindR/full/early_vs_mature_markers.csv"
                  )

# Load all relevant marker files
if (all(file.exists(marker_files))) {
  
  early_vs_inter_markers <- read_csv(marker_files[1])
  inter_vs_trans_markers <- read_csv(marker_files[2])
  trans_vs_mature_markers <- read_csv(marker_files[3])
  early_vs_mature_markers <- read_csv(marker_files[4])

} else {
  
  stop("One or more marker files are missing. Run FindMarkers first.")
  
}

# Make function to reformat to pathfindR input
reformat_pathfindR <- function(input) {
  
  input %>%
    dplyr::select(Gene.symbol = ...1, logFC = avg_log2FC, adj.P.Val = p_val_adj) %>% 
    as.data.frame()

}

early_vs_inter_markers <- reformat_pathfindR(early_vs_inter_markers)
inter_vs_trans_markers <- reformat_pathfindR(inter_vs_trans_markers)
trans_vs_mature_markers <- reformat_pathfindR(trans_vs_mature_markers)
early_vs_mature_markers <- reformat_pathfindR(early_vs_mature_markers)

```


```{r  fig.width=16, fig.height=6}
term_gene_heatmap(step1_output, use_description = TRUE, sort_terms_by_p = TRUE)
```


```{r  fig.width=16, fig.height=6}
term_gene_heatmap(step2_output, use_description = TRUE, sort_terms_by_p = TRUE)
```


```{r  fig.width=16, fig.height=6}
term_gene_heatmap(step3_output, use_description = TRUE, sort_terms_by_p = TRUE)
```


```{r}
term_gene_heatmap(full_diff_output, use_description = TRUE, sort_terms_by_p = TRUE, num_terms = 20, genes_df = example_pathfindR_input))
```

```{r}

UpSet_plot(full_diff_output, use_description = TRUE)

```


```{r}
enrichment_chart(step1_output)
```



## chat attempt

```{r}

library(simplifyEnrichment)

# Provide GO IDs
go_ids <- step1_output$ID

# Compute semantic similarity matrix
sim_mat <- GO_similarity(go_ids)

# Convert similarity matrix to distance matrix
dist_mat <- as.dist(1 - sim_mat)

# Run base R hierarchical clustering
hc <- hclust(dist_mat, method = "ward.D2")

# Cut tree into desired number of clusters
clusters <- cutree(hc, k = 5)

# Visualise using simplifyEnrichment with manual clustering
simplifyEnrichment(sim_mat, cluster_by = clusters)

```




## pathfindR

### Define function

Cluster enriched terms. This will group enriched pathways into clusters, and each cluster will be given a certain number of "representative" pathways. It is common practice to filter this output as follows:
- Fold enrichment > 2: this is considered biologically meaningful, as the term is present at least twice as much as in the background
- Occurrence >= 9: this ensures only terms present in over 8 out of 10 `pathfindR` iterations are kept, for high confidence
- Lowest_p < 0.05: this ensures only statistically significant terms are kept

This function `cluster_and_plot_pathways` filters representative pathways based on robustness criteria, visualises them as a bar plot, and saves the result.

```{r fig.width=11, fig.height=7}

cluster_and_plot_pathways <- function(pathfindR_output, output_prefix, title_label = NULL,
                                      output_dir = "results/humanPVATsn/pathfindR/full/",
                                      save_plot = TRUE, save_csv = TRUE) {
  
  # Cluster enriched terms
  clustered <- cluster_enriched_terms(pathfindR_output, plot_dend = FALSE, plot_clusters_graph = FALSE)
  
  # Filter to keep only robust representative terms
  clustered_rep <- clustered %>%
    filter(Status == "Representative",
           Fold_Enrichment > 2,
           occurrence >= 8,
           lowest_p < 0.05)
  
  # Handle title if not provided
  if (is.null(title_label)) {
    title_label <- paste("Clusters of enriched pathways –", output_prefix)
  }

  # Assign colour based on step
  fill_col <- case_when(
    output_prefix == "step1" ~ "#FDD866",
    output_prefix == "step2" ~ "#F46D43",
    output_prefix == "step3" ~ "#D73027",
    output_prefix %in% c("full_diff", "full_diff_inv") ~ "#2166AC"
  )
  
  # Visualise with a barplot
  plot <- ggplot(clustered_rep,
               aes(x = Fold_Enrichment, y = reorder(Term_Description, Fold_Enrichment))) +
          geom_bar(stat = "identity", color = "black", fill = fill_col) +
          labs(title = title_label,
               x = "Fold enrichment",
               y = "Pathway") +
          theme_minimal() +
          theme(legend.position = "none", plot.margin = margin(10, 20, 10, 10))
  
  # Define output file paths
  pdf_file <- file.path(output_dir, paste0(output_prefix, "_clustered.pdf"))
  csv_file <- file.path(output_dir, paste0(output_prefix, "_clustered_filtered.csv"))
  
  # Save outputs
  if (save_plot) ggsave(pdf_file, plot, width = 13, height = 6, device = "pdf")
  if (save_csv) write.csv(clustered_rep, csv_file, row.names = FALSE)

  # Return table and plot
  return(list(data = clustered_rep, plot = plot))

}

```

### Call function

This function is called on pathway enrichment analyses for each step of differentiation and also one on the full differentiation process.

```{r fig.width=13, fig.height=6}

# Define inputs
pathfindR_outputs <- list(
  step1 = step1_output,
  step2 = step2_output,
  step3 = step3_output,
  full_diff = full_diff_output
)

# Define matching labels and output names
run_labels <- list(
  step1 = "Step 1: enriched pathway clusters across the early-to-intermediate adipocyte transition",
  step2 = "Step 2: enriched pathway clusters across the intermediate-to-transitional adipocyte transition",
  step3 = "Step 3: enriched pathway clusters across the transitional-to-mature adipocyte transition",
  full_diff = "Full differentiation: enriched pathway clusters across the early-to-mature adipocyte transition"
)

# Run all analyses
results_list <- list()

for (name in names(pathfindR_outputs)) {
  
  # Call the function on each adipocyte stage
  output <- cluster_and_plot_pathways(
    pathfindR_output = pathfindR_outputs[[name]],
    output_prefix = paste0(name),
    title_label = run_labels[[name]]
  )
  print(paste0(name))
  
  # Save results
  results_list[[name]] <- output
  
  # Print the plots
  print(output$plot)

}

```

# Final plots

These contain the umbrella terms as y-axis titles and the heights of the facet wrapped plots are proportional to the number of pathways the umbrella term contains.

Load data

```{r}

step1_output <- readRDS("results/humanPVATsn/pathfindR/full/step1.rds")
step2_output <- readRDS("results/humanPVATsn/pathfindR/full/step2.rds")
step3_output <- readRDS("results/humanPVATsn/pathfindR/full/step3.rds")
full_diff <- readRDS("results/humanPVATsn/pathfindR/full/full_diff.rds")

```


## Step 1

```{r fig.width=10, fig.height=9}

# Filter pathfindR output
step1_filtered <- step1_output %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only significantly enriched pathways
  filter(lowest_p < 0.05) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

# Biological theme list
theme_list <- c(
  "Cellular & tissue \nremodelling",
  "Cell signalling",
  "Neurogenic \npriming",
  "Metabolic priming \n& gene regulation"
)

# Assign themes
step1_filtered$Theme <- c(
  theme_list[2],  # Roundabout signaling pathway
  theme_list[3],  # positive regulation of axonogenesis
  theme_list[3],  # axon extension involved in axon guidance
  theme_list[1],  # negative regulation of cardiac muscle hypertrophy
  theme_list[3],  # dentate gyrus development
  theme_list[2],  # interleukin-11-mediated signaling pathway
  theme_list[4],  # aspartate transmembrane transport
  theme_list[1],  # regulation of cell junction assembly
  theme_list[1],  # positive regulation of T cell migration
  theme_list[1],  # negative chemotaxis
  theme_list[1],  # negative regulation of cell size
  theme_list[1],  # apical junction assembly
  theme_list[1],  # pulmonary valve morphogenesis
  theme_list[4],  # positive regulation of glycolytic process
  theme_list[2],  # calcineurin-NFAT signaling cascade
  theme_list[1],  # substrate adhesion-dependent cell spreading
  theme_list[4],  # P-body assembly
  theme_list[4],  # heterochromatin organization
  theme_list[1],  # regulation of Cdc42 protein signal transduction
  theme_list[1]   # positive regulation of stress fiber assembly
)

# Set Theme factor levels to control facet order
step1_filtered$Theme <- factor(step1_filtered$Theme, levels = theme_list)

# Reorder by theme and pathway name
step1_filtered <- step1_filtered %>%
  arrange(Theme, Term_Description) %>%
  mutate(Term_Description = factor(Term_Description, levels = Term_Description))

# Split data by Theme
data_by_theme <- split(step1_filtered, step1_filtered$Theme)

# Compute relative heights based on pathway counts
relative_heights <- sapply(data_by_theme, nrow)

# Max value for consistent x-axis
max_value <- max(step1_filtered$Fold_Enrichment)

# Generate dot plots for each theme with theme label as y-axis title
plots_by_theme <- lapply(names(data_by_theme), function(theme) {
  df <- data_by_theme[[theme]]
  ggplot(df, aes(x = Fold_Enrichment, y = Term_Description)) +
    geom_point(aes(size = occurrence), colour = "#FDD866", stroke = 0.3, alpha = 0.8) +
    scale_size_continuous(
      name = "Occurrence",
      range = c(3, 6),
      breaks = c(6, 8, 10),
      limits = c(6, 10),
      labels = c("6", "8", "10"),
      guide = guide_legend(override.aes = list(size = c(3, 4.5, 6)))) +
    xlim(0, max_value) +
    labs(x = NULL, y = theme) +
    theme_minimal() +
    theme(
      axis.title.y = element_text(angle = 90, hjust = 0.5, face = "bold", size = 13),
      axis.text.y = element_text(size = 13),
      legend.position = "right",
      legend.title = element_text(size = 13),
      legend.text = element_text(size = 12),
      axis.text.x = element_text(size = 12),
      plot.margin = margin(5, 5, 5, 5)
    )
})

# Combine plots with proportional heights
combined_plot <- wrap_plots(plots_by_theme, ncol = 1, heights = relative_heights) +
  plot_layout(guides = 'collect', heights = relative_heights + 0.5, ncol = 1) & 
  theme(plot.margin = margin(10, 10, 20, 10))

# Add global title and x-axis
final_plot <- combined_plot + plot_annotation(
  caption = "Fold enrichment",
  theme = theme(
    plot.caption = element_text(hjust = 0.8, size = 12))
)

# Save and display
ggsave("results/humanPVATsn/final_figs/step1.pdf", final_plot, width = 10, height = 9)
print(final_plot)

```


## Step 2

```{r fig.width=10, fig.height=9}

# Filter pathfindR output
step2_filtered <- step2_output %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only significantly enriched pathways
  filter(lowest_p < 0.05) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

# Biological theme list
theme_list <- c(
  "Metabolic \npriming",
  "Cell signalling & \ntissue remodelling",
  "Neurogenic \npriming",
  "Lineage \norigins",
  "Amyloid-\n beta \n processing"
)


# Assign themes
step2_filtered$Theme <- c(
  theme_list[1],
  theme_list[1],
  theme_list[4],
  theme_list[1],
  theme_list[2],
  theme_list[1],
  theme_list[2],
  theme_list[2],
  theme_list[4],
  theme_list[1],
  theme_list[1],
  theme_list[1],
  theme_list[1],
  theme_list[5],
  theme_list[1],
  theme_list[2],
  theme_list[3],
  theme_list[3],
  theme_list[3],
  theme_list[1]
)

# Set Theme factor levels to control facet order
step2_filtered$Theme <- factor(step2_filtered$Theme, levels = theme_list)

# Reorder by theme and pathway name
step2_filtered <- step2_filtered %>%
  arrange(Theme, Term_Description) %>%
  mutate(Term_Description = factor(Term_Description, levels = Term_Description))

# Split data by Theme
data_by_theme <- split(step2_filtered, step2_filtered$Theme)

# Compute relative heights based on pathway counts
relative_heights <- sapply(data_by_theme, nrow)

# Generate plots for each theme with theme label as y-axis title
max_value <- max(step2_filtered$Fold_Enrichment)

plots_by_theme <- lapply(names(data_by_theme), function(theme) {
  df <- data_by_theme[[theme]]
  ggplot(df, aes(x = Fold_Enrichment, y = Term_Description)) +
    geom_point(aes(size = occurrence), colour = "#F46D43", stroke = 0.3, alpha = 0.8) +
    scale_size_continuous(
      name = "Occurrence",
      range = c(3, 6),
      breaks = c(6, 8, 10),
      limits = c(6, 10),
      labels = c("6", "8", "10"),
      guide = guide_legend(override.aes = list(size = c(3, 4.5, 6)))) +
    xlim(0, max_value) +
    labs(
      x = NULL,
      y = if (theme == "Amyloid-\n beta \n processing") {
        bquote(bold(atop("Amyloid-"*beta, "processing"))) } else { theme }
      ) +
    theme_minimal() +
    theme(
      axis.title.y = element_text(angle = 90, hjust = 0.5, face = "bold", size = 13),
      axis.text.y = element_text(size = 13),
      legend.position = "right",
      legend.title = element_text(size = 13),
      legend.text = element_text(size = 12),
      axis.text.x = element_text(size = 12),
      plot.margin = margin(5, 5, 5, 5)
    )
})

# Combine plots with proportional heights
combined_plot <- wrap_plots(plots_by_theme, ncol = 1, heights = relative_heights) +
  plot_layout(guides = 'collect', heights = relative_heights + 0.5, ncol = 1) & 
  theme(plot.margin = margin(10, 10, 20, 10))

# Add global title and x-axis
final_plot <- combined_plot + plot_annotation(
  caption = "Fold enrichment",
  theme = theme(
    plot.caption = element_text(hjust = 0.8, size = 12))
)

# Save and display
ggsave("results/humanPVATsn/final_figs/step2.pdf", final_plot, width = 10, height = 9)
print(final_plot)


```


## Step 3

```{r fig.width=9, fig.height=7}

# Filter pathfindR output
step3_filtered <- step3_output %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only significantly enriched pathways
  filter(lowest_p < 0.05) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

# Biological theme list
theme_list <- c(
  "Metabolic \nmaturation",
  "Cell & tissue \nremodelling",
  "Cell signalling",
  "Sympathetic \ninteraction",
  "Amyloid-beta \nprocessing"
)

# Assign themes
step3_filtered$Theme <- c(
  theme_list[3],  # negative regulation of cardiac muscle hypertrophy
  theme_list[1],  # intracellular distribution of mitochondria
  theme_list[2],  # sequestering of actin monomers
  theme_list[1],  # white fat cell differentiation
  theme_list[4],  # axon extension involved in axon guidance
  theme_list[4],  # synaptic membrane adhesion
  theme_list[4],  # negative regulation of excitatory postsynaptic potential
  theme_list[2],  # regulation of vesicle size
  theme_list[3],  # lncRNA-mediated post-transcriptional gene silencing
  theme_list[2],  # establishment of epithelial cell polarity
  theme_list[4],  # hindbrain development
  theme_list[2],  # endoplasmic reticulum tubular network organisation
  theme_list[4],  # locomotor rhythm
  theme_list[3],  # regulation of smooth muscle cell proliferation
  theme_list[3],  # cellular response to fibroblast growth factor stimulus
  theme_list[3],  # positive regulation of phospholipase C activity
  theme_list[4],  # dendritic spine maintenance
  theme_list[5],  # amyloid-beta clearance
  theme_list[3],  # positive regulation of miRNA catabolic process
  theme_list[4]   # astrocyte activation
)

# Set Theme factor levels to control facet order
step3_filtered$Theme <- factor(step3_filtered$Theme, levels = theme_list)

# Reorder by theme and pathway name
step3_filtered <- step3_filtered %>%
  arrange(Theme, Term_Description) %>%
  mutate(Term_Description = factor(Term_Description, levels = Term_Description))

# Split data by Theme
data_by_theme <- split(step3_filtered, step3_filtered$Theme)

# Compute relative heights based on pathway counts
relative_heights <- sapply(data_by_theme, nrow)

# Generate plots for each theme with theme label as y-axis title
max_value <- max(step3_filtered$Fold_Enrichment)

plots_by_theme <- lapply(names(data_by_theme), function(theme) {
  df <- data_by_theme[[theme]]
  ggplot(df, aes(x = Fold_Enrichment, y = Term_Description)) +
    geom_point(aes(size = occurrence), colour = "#D73027", stroke = 0.3, alpha = 0.8) +
    scale_size_continuous(
      name = "Occurrence",
      range = c(3, 6),
      breaks = c(6, 8, 10),
      limits = c(6, 10),
      labels = c("6", "8", "10"),
      guide = guide_legend(override.aes = list(size = c(3, 4.5, 6)))) +
    xlim(0, max_value) +
    labs(
      x = NULL,
      y = if (theme == "Amyloid-beta \nprocessing") {
        bquote(bold(atop("Amyloid-"*beta, "processing"))) } else { theme }
    ) +
    theme_minimal() +
    theme(
      axis.title.y = element_text(angle = 90, hjust = 0.5, face = "bold", size = 13),
      axis.text.y = element_text(size = 13),
      legend.position = "right",
      legend.title = element_text(size = 13),
      legend.text = element_text(size = 12),
      axis.text.x = element_text(size = 12),
      plot.margin = margin(5, 5, 5, 5)
    )
})

# Combine plots with proportional heights
combined_plot <- wrap_plots(plots_by_theme, ncol = 1, heights = relative_heights) +
  plot_layout(guides = 'collect', heights = relative_heights + 0.5, ncol = 1) & 
  theme(plot.margin = margin(10, 10, 20, 10))

# Add global title and x-axis
final_plot <- combined_plot + plot_annotation(
  caption = "Fold enrichment",
  theme = theme(
    plot.caption = element_text(hjust = 0.8, size = 12))
)

# Save and display
ggsave("results/humanPVATsn/final_figs/step3.pdf", final_plot, width = 10, height = 9)
print(final_plot)

```


## Full differentiation

```{r fig.width=12, fig.height=9}

# Filter pathfindR output
full_diff_filtered <- full_diff %>%
  # Sort by descending fold enrichment
  arrange(desc(.data[["Fold_Enrichment"]])) %>%
  # Keep only significantly enriched pathways
  filter(lowest_p < 0.05) %>%
  # Keep only those with more than 5 occurrences
  filter(occurrence > 5) %>%
  # Keep top 20 enriched pathways
  slice_head(n = 20) %>%
  # Ensure factor levels are correctly ordered for plotting
  mutate(Term_Description = factor(Term_Description,
                                   levels = rev(Term_Description)))

# Biological theme list
theme_list <- c(
  "Metabolic activation",
  "Cellular restructuring \n& vesicle dynamics",
  "Cell communication \n& signal transduction",
  "Differentiation \n& tissue remodelling"
)

# Assign themes
full_diff_filtered$Theme <- c(
  theme_list[3],
  theme_list[1],
  theme_list[1],
  theme_list[2],
  theme_list[1],
  theme_list[2],
  theme_list[1],
  theme_list[4],
  theme_list[4],
  theme_list[2],
  theme_list[2],
  theme_list[2],
  theme_list[1],
  theme_list[3],
  theme_list[3],
  theme_list[3],
  theme_list[4],
  theme_list[3],
  theme_list[1],
  theme_list[4]
)

# Set Theme factor levels to control facet order
full_diff_filtered$Theme <- factor(full_diff_filtered$Theme, levels = theme_list)

# Reorder by theme and pathway name
full_diff_filtered <- full_diff_filtered %>%
  arrange(Theme, Term_Description) %>%
  mutate(Term_Description = factor(Term_Description, levels = Term_Description))

# Split data by Theme
data_by_theme <- split(full_diff_filtered, full_diff_filtered$Theme)

# Compute relative heights based on pathway counts
relative_heights <- sapply(data_by_theme, nrow)

# Generate plots for each theme with theme label as y-axis title
max_value <- max(full_diff_filtered$Fold_Enrichment)

plots_by_theme <- lapply(names(data_by_theme), function(theme) {
  df <- data_by_theme[[theme]]
  ggplot(df, aes(x = Fold_Enrichment, y = Term_Description)) +
    geom_point(aes(size = occurrence), colour = "#2166AC", stroke = 0.3, alpha = 0.8) +
    scale_size_continuous(
      name = "Occurrence",
      range = c(3, 6),
      breaks = c(6, 8, 10),
      limits = c(6, 10),
      labels = c("6", "8", "10"),
      guide = guide_legend(override.aes = list(size = c(3, 4.5, 6)))) +
    xlim(0, max_value) +
    labs(
      x = NULL,
      y = if (theme == "Amyloid-beta \nprocessing") {
        bquote(bold(atop("Amyloid-"*beta, "processing"))) } else { theme }
    ) +
    theme_minimal() +
    theme(
      axis.title.y = element_text(angle = 90, hjust = 0.5, face = "bold", size = 13),
      axis.text.y = element_text(size = 13),
      legend.position = "right",
      legend.title = element_text(size = 13),
      legend.text = element_text(size = 12),
      axis.text.x = element_text(size = 12),
      plot.margin = margin(5, 5, 5, 5)
    )
})

# Combine plots with proportional heights
combined_plot <- wrap_plots(plots_by_theme, ncol = 1, heights = relative_heights) +
  plot_layout(guides = 'collect', heights = relative_heights + 0.5, ncol = 1) & 
  theme(plot.margin = margin(10, 10, 20, 10))

# Add global title and x-axis
final_plot <- combined_plot + plot_annotation(
  caption = "Fold enrichment",
  theme = theme(
    plot.caption = element_text(hjust = 0.8, size = 12))
)

# Save and display
ggsave("results/humanPVATsn/final_figs/full_diff.pdf", final_plot, width = 10, height = 9)
print(final_plot)

```

# Session info

```{r}

sessionInfo()

```

