---
title: "Gene set enrichment analysis on snRNA-sequencing samples of human perivascular adipose tissue (PVAT)"
author: "Hannah M. L. Bazin"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
---

In this vignette I run a gene set enrichment analysis (GSEA) on four clusters of brown adipocyte lineage. The clusters have been identified using functions from Seurat 4.4.0 in the notebook named `human_PVAT_snRNAseq.Rmd`. I am building on code from Holly A. R. Giles.


# Set up

## Set working directory

The working directory can be set by running the following command in the terminal: `setwd("/Users/hannahbazin/Desktop/Cambridge/Academics/Han_Lab/MPhil/mphil-project")`


## Set seed


```{r}

# Set random seed for reproducibility
set.seed(1996)

```

## Load libraries

```{r}

# Load software packages 
library('org.Hs.eg.db')
library(msigdbr)
library(pheatmap)
library("clusterProfiler")
library(Seurat)
library(tidyverse)

```


## Load data

To save on time we reload the integrated Seurat object generated in `human_PVAT_snRNAseq.Rmd`.

```{r}

# Load integrated Seurat object from previous analysis
load(file = "data/analysis/anno_combined.RData")

```


# Pre-processing workflow

## Subset adipocyte clusters

Since the aim is to identify differentially activated pathways in the four adipocyte-related clusters, we subset the Seurat object to contain only these clusters.

```{r}

# Select clusters in adipocyte lineage
adipo_clusters <- c("Early pre-adipocytes",
                    "Intermediate pre-adipocytes",
                    "Transitional adipocytes",
                    "Mature adipocytes")

# Subset Seurat object to include only adipocyte clusters
anno_combined_adipo <- subset(anno_combined, idents = adipo_clusters)

```

## Scale data

Subsetting a Seurat object may cause the loss of the scaled data in the object. We verify this and recompute the scaled data as needed. This is a time-consuming step, so it is best to save the R object.

```{r}

# Ensure default assay is RNA assay
DefaultAssay(anno_combined_adipo) <- "RNA"

# Check that scaled data was preserved - this can be lost while subsetting
head(anno_combined_adipo@assays$RNA@scale.data) # NOT OK

# Scale data if needed
  # This is a time-consuming step, save the object
if (file.exists("data/analysis/anno_combined_adipo_scaled.RData")) {
  
  message("Loading existing scaled data file.")
  
  load(file = "data/analysis/anno_combined_adipo_scaled.RData")
  
} else {
  
  message("Missing scaled data file. Scaling data now.")
  
  anno_combined_adipo_scaled <- ScaleData(anno_combined_adipo, vars.to.regress = "percent.mt")
  
  save(anno_combined_adipo_scaled, file = "data/analysis/anno_combined_adipo_scaled.RData")

}

```

# Find markers

The Seurat `FindMarkers` function can identify differentially expressed genes by comparing one cluster with another cluster. To rerun the full analysis from scratch, be sure to delete any existing marker files in the folder `results/humanPVATsn/GSEA/`.

We run `FindMarkers()` once per differentiation step, and also once for the full differentiation. The parameters `logfc.threshold = -Inf, min.pct = -Inf, min.diff.pct = -Inf` ensure that all genes are kept in the dataframe, not just the differentially expressed ones. This is necessary for GSEA.

### Set up

```{r}

# Ensure default assay is RNA assay
DefaultAssay(anno_combined_adipo_scaled) <- "RNA"

marker_files <- c("results/humanPVATsn/GSEA/early_vs_inter_markers.csv",
                  "results/humanPVATsn/GSEA/inter_vs_trans_markers.csv",
                  "results/humanPVATsn/GSEA/trans_vs_mature_markers.csv",
                  "results/humanPVATsn/GSEA/early_vs_mature_markers.csv"
                  )

```

### Early vs intermediate pre-adipocytes

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[1])) {
  
  early_vs_inter_markers <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Early pre-adipocytes",
                                      ident.2 = "Intermediate pre-adipocytes",
                                      logfc.threshold = -Inf,
                                      min.pct = -Inf,
                                      min.diff.pct = -Inf
                                      )
  
  write.csv(early_vs_inter_markers, marker_files[1], row.names = TRUE)
  
  message(paste0("Saved markers for early vs intermediate pre-adipocytes to ", marker_files[1]))
  
} else {
  
  message("Marker file for early vs intermediate pre-adipocytes already exists. Skipping FindMarkers.")
  
}

```

### Intermediate pre-adipocytes vs transitional adipocytes

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[2])) {
  
  inter_vs_trans_markers <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Intermediate pre-adipocytes",
                                      ident.2 = "Transitional adipocytes",
                                      logfc.threshold = -Inf,
                                      min.pct = -Inf,
                                      min.diff.pct = -Inf
                                      )
  
  write.csv(inter_vs_trans_markers, marker_files[2], row.names = TRUE)
  
  message(paste0("Saved markers for intermediate pre-adipocytes vs transitional adipocytes to ", marker_files[2]))
  
} else {
  
  message("Marker file for intermediate pre-adipocytes vs transitional adipocytes already exists. Skipping FindMarkers.")
  
}

```

### Transitional adipocytes vs mature adipocytes

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[3])) {
  
  trans_vs_mature_markers <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Transitional adipocytes",
                                      ident.2 = "Mature adipocytes",
                                      logfc.threshold = -Inf,
                                      min.pct = -Inf,
                                      min.diff.pct = -Inf
                                      )
  
  write.csv(trans_vs_mature_markers, marker_files[3], row.names = TRUE)
  
  message(paste0("Saved markers for transitional vs mature adipocytes to ", marker_files[3]))
  
} else {
  
  message("Marker file for transitional vs mature adipocytes already exists. Skipping FindMarkers.")
  
}

```

### Early pre-adipocytes vs mature adipocytes

```{r}

# Find differentially expressed genes
  # This is a time-consuming step, save the object
if (!file.exists(marker_files[4])) {
  
  early_vs_mature_markers <- FindMarkers(object = anno_combined_adipo_scaled,
                                      ident.1 = "Early pre-adipocytes",
                                      ident.2 = "Mature adipocytes",
                                      logfc.threshold = -Inf,
                                      min.pct = -Inf,
                                      min.diff.pct = -Inf
                                      )
  
  write.csv(early_vs_mature_markers, marker_files[4], row.names = TRUE)
  
  message(paste0("Saved markers for early pre-adipocytes vs mature adipocytes to ", marker_files[4]))
  
} else {
  
  message("Marker file for early pre-adipocytes vs mature adipocytes already exists. Skipping FindMarkers.")
  
}

```


# GSEA

## Load marker files

To streamline the workflow we save the marker files as `csv` files in the folder `results/humanPVATsn/GSEA/`. These marker files can be loaded into RStudio to save the user from running FindMarkers() again, which is a time-consuming step.

```{r}

# Redefine marker files if starting code from this chunk
marker_files <- c("results/humanPVATsn/GSEA/early_vs_inter_markers.csv",
                  "results/humanPVATsn/GSEA/inter_vs_trans_markers.csv",
                  "results/humanPVATsn/GSEA/trans_vs_mature_markers.csv",
                  "results/humanPVATsn/GSEA/early_vs_mature_markers.csv"
                  )

# Load all relevant marker files
if (all(file.exists(marker_files))) {
  
  early_vs_inter_markers <- read_csv(marker_files[1])
  inter_vs_trans_markers <- read_csv(marker_files[2])
  trans_vs_mature_markers <- read_csv(marker_files[3])
  early_vs_mature_markers <- read_csv(marker_files[4])
  
} else {
  
  stop("One or more marker files are missing. Run FindMarkers first.")
  
}

```


## Reformat data

GSEA requires a ranked gene list, where genes are ranked based on some statistical measure (e.g., log fold change, Wald statistic, or p-value). I will rank the genes based on log fold change (`avg_log2FC`).


```{r}

# Select relevant columns for GSEA input
step1_scores <- early_vs_inter_markers %>% 
  rownames_to_column(var = "gene") %>% 
  select(gene, avg_log2FC)

step2_scores <- inter_vs_trans_markers %>% 
  rownames_to_column(var = "gene") %>% 
  select(gene, avg_log2FC)

step3_scores <- trans_vs_mature_markers %>% 
  rownames_to_column(var = "gene") %>% 
  select(gene, avg_log2FC)

full_diff_scores <- early_vs_mature_markers %>% 
  rownames_to_column(var = "gene") %>% 
  select(gene, avg_log2FC)

```






















