---
title: "Pathway enrichment analysis of human perivascular adipose tissue (PVAT) snRNA-seq data"
author: "Hannah M. L. Bazin"
date: "`r doc_date()`"
output:
  BiocStyle::html_document:
      toc: yes
      toc_depth: 3
      toc_float: yes
---

In this vignette, I run pathfindR on gene markers identified in the analysis of single-nucleus RNA sequencing of human perivascular adipose tissue (PVAT). The markers were generated using Seurat's FindAllMarkers() function; the full code can be found in the vignette `human_PVAT_snRNAseq.Rmd`. The raw data is available on the Gene Expression Omnibus website [at this link](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE164528).

# Overview of `pathfindR` algorithm

`pathfindR` is an R package designed for active-subnetwork-oriented pathway enrichment analysis. Unlike traditional over-representation analysis (ORA) and gene set enrichment analysis (GSEA), which evaluate gene lists without considering gene interactions, pathfindR integrates protein-protein interaction networks (PINs) to identify active subnetworks – groups of interacting genes enriched in significantly altered genes. A subnetwork is defined as a cluster of interconnected genes in a PIN, and it is considered active if it contains a disproportionately high number of differentially expressed genes. The algorithm then performs pathway enrichment analysis on these subnetworks to identify biologically relevant pathways.

The `pathfindR` workflow consists of the following steps:

1.  **Mapping input genes** and their p-values onto a predefined PIN.
2.  **Identification of active subnetworks**, using a heuristic search algorithm to detect interconnected gene clusters enriched in significant genes.
3.  **Filtering subnetworks** based on predefined scoring criteria, including (i) the number of significant genes (minimum of 3 by default) and (ii) the subnetwork score, calculated as the sum of absolute log-transformed p-values of significant genes.
4.  **Pathway enrichment analysis** on selected subnetworks using Fisher’s Exact Test, with multiple testing correction using the Bonferroni method by default.
5.  **Iterative analysis**, where the active subnetwork search and enrichment steps are repeated multiple times to account for random variation in subnetwork detection, ensuring robustness by identifying pathways that remain significantly enriched across multiple runs.
6.  **Output generation**, producing a structured data frame of significantly enriched pathways and gene sets, along with key metrics such as the lowest and highest adjusted p-values across iterations, to assess the reproducibility of pathway enrichment across iterations.

# Set up

## Set working directory

The working directory can be set by running the following command in the terminal: `setwd("/Users/hannahbazin/Desktop/Cambridge/Academics/Han_Lab/MPhil/mphil-project")`

## Load libraries

```{r}

library(msigdbr)
library(pathfindR)
library(rJava)
library(pathfindR)
library(tidyverse)

```

## Set seed

Because `pathfindR` includes a heuristic search algorithm to identify active subnetworks, we set a random seed for reproducibility.

```{r}

set.seed(1996)

```


## Load data

We load in the markers for clusters of interest: 1, 6, 9, and 0. These all correspond to adipocyte lineage, while the other clusters do not.

```{r}

mature <- read_csv("results/humanPVATsn/markersCluster0.csv")
transitional <- read_csv("results/humanPVATsn/markersCluster9.csv")
intermediate <- read_csv("results/humanPVATsn/markersCluster1.csv")
early <- read_csv("results/humanPVATsn/markersCluster6.csv")

```


# Process data

The data needs to be reformatted to align with the expected input of `pathfindR`.

```{r}

early <- early %>% select(Gene.symbol = gene, logFC = avg_log2FC, adj.P.Val = p_val_adj)

intermediate <- intermediate %>% select(Gene.symbol = gene, logFC = avg_log2FC, adj.P.Val = p_val_adj)

mature <- mature %>% select(Gene.symbol = gene, logFC = avg_log2FC, adj.P.Val = p_val_adj)

transitional <- transitional %>% select(Gene.symbol = gene, logFC = avg_log2FC, adj.P.Val = p_val_adj)


```


# Run analysis

## Import gene sets

Since `pathfindR` does not natively support Molecular Signature Database (MSigDB) gene sets, we must import them.

```{r}

c2_gene_sets <- msigdbr(species = "Homo sapiens", category = "C2")

```


## Run `pathfindR`

We run the analysis separately for each cluster (i.e. step of adipocyte differentiation). This will highlight which pathways are active in each developmental stage. For each cluster, we run the analysis against four different gene sets.

```{r}

# Define list of clusters
clusters <- list(
  "early" = early,
  "intermediate" = intermediate,
  "transitional" = transitional,
  "mature" = mature
)

# Define pathway datasets ??
pathway_sets <- c("C2", "C2:CGP", "C2:REACTOME", "C5:GOBP")

# Loop through each cluster and pathway dataset
for (cluster_name in names(clusters)) {
  
  for (gene_set in pathway_sets) {
    
    # Run pathfindR
    output <- run_pathfindR(
      clusters[[cluster_name]],
      gene_sets = gene_set,
      min_gset_size = 5, # Include smaller adipocyte-related pathways
      max_gset_size = 500,
      adj_method = "fdr",
      enrichment_threshold = 0.01,
      iterations = 10, # Increase if needed for robustness
      output_dir = paste0("results/humanPVATsn/pathfindR/", cluster_name, "_", gsub(":", "_", gene_set))
    )
    
    # Save results
    saveRDS(output, paste0("results/humanPVATsn/pathfindR/pathfindR_", cluster_name, "_", gsub(":", "_", gene_set), ".rds"))
  }
}
                        

```













